nano/node/active_transactions.cpp

```diff
@@ -770,37 +770,85 @@ std::unique_ptr<nano::container_info_component> nano::recently_confirmed_cache::
  */
 
 nano::recently_cemented_cache::recently_cemented_cache (std::size_t max_size_a) :
-	max_size{ max_size_a }
+	max_size{ max_size_a },
+	handle (rsnano::rsn_recently_cemented_cache_create1 (max_size_a))
+{
+}
+
+nano::recently_cemented_cache::recently_cemented_cache (nano::recently_cemented_cache const & other_a) :
+	max_size{ other_a.max_size },
+	handle (rsnano::rsn_recently_cemented_cache_clone (other_a.handle))
 {
 }
 
+nano::recently_cemented_cache::~recently_cemented_cache ()
+{
+	if (handle != nullptr)
+		rsnano::rsn_recently_cemented_cache_destroy (handle);
+}
+
+nano::recently_cemented_cache & nano::recently_cemented_cache::operator= (const nano::recently_cemented_cache & other_a)
+{
+	if (handle != nullptr)
+		rsnano::rsn_recently_cemented_cache_destroy (handle);
+
+	handle = rsnano::rsn_recently_cemented_cache_clone (other_a.handle);
+	return *this;
+}
+
 void nano::recently_cemented_cache::put (const nano::election_status & status)
 {
 	nano::lock_guard<nano::mutex> guard{ mutex };
-	cemented.push_back (status);
-	if (cemented.size () > max_size)
-	{
-		cemented.pop_front ();
-	}
+	rsnano::rsn_recently_cemented_cache_put (handle, status.handle);
 }
 
 nano::recently_cemented_cache::queue_t nano::recently_cemented_cache::list () const
 {
-	nano::lock_guard<nano::mutex> guard{ mutex };
-	return cemented;
+	rsnano::RecentlyCementedCachedDto recently_cemented_cache_dto;
+	rsnano::rsn_recently_cemented_cache_list (handle, &recently_cemented_cache_dto);
+	nano::recently_cemented_cache::queue_t result;
+	rsnano::RecentlyCementedCacheItemDto const * current;
+	int i;
+	for (i = 0, current = recently_cemented_cache_dto.items; i < recently_cemented_cache_dto.count; ++i)
+	{
+		std::shared_ptr<nano::block> const & winner = nano::block_handle_to_block (current->winner);
+		nano::uint128_t tally;
+		boost::multiprecision::import_bits (tally, std::begin (current->tally), std::end (current->tally), 8, true);
+		nano::uint128_t final_tally;
+		boost::multiprecision::import_bits (final_tally, std::begin (current->final_tally), std::end (current->final_tally), 8, true);
+		nano::election_status election_status { winner };
+		election_status.set_tally (tally);
+		election_status.set_final_tally (final_tally);
+		election_status.set_confirmation_request_count (current->confirmation_request_count);
+		election_status.set_block_count (current->block_count);
+		election_status.set_voter_count (current->voter_count);
+		election_status.set_election_end (std::chrono::milliseconds (current->election_end));
+		election_status.set_election_duration (std::chrono::milliseconds (current->election_duration));
+		election_status.set_election_status_type (static_cast<nano::election_status_type> (current->election_status_type));
+		result.push_back (election_status);
```

Comment:
The problem I am having with this approach is that ``ElectionStatusHandle`` is an incomplete pointer and does not know the size of the ``ElectionStatus``object at compile time. Because of that I can't use arithmetic on the pointer and do ``current++``on the loop that converts the ``ElectionStatusHandle``into ``nano::election_status``. 

Any idea how this issue can be solved?

Reply:
Oh yes, I didn't think of that. So current should be a pointer to a pointer. and the Rust DTO struct should probably look like this then:
```
#[repr(C)]
pub struct RecentlyCementedCachedDto {
    items: *const *mut ElectionStatusHandle,
    count: usize,
    pub raw_data: *mut RecentlyCementedCachedRawData,
}```